libclc/generic/lib/math/clc_ldexp.cl

```diff
@@ -0,0 +1,128 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include <clc/clc.h>
+#include "config.h"
+#include "../clcmacro.h"
+#include "math.h"
+
+_CLC_DEF _CLC_OVERLOAD float __clc_ldexp(float x, int n) {
+
+	if (!__clc_fp32_subnormals_supported()) {
+
+		// This treats subnormals as zeros
+		int i = as_int(x);
+		int e = (i >> 23) & 0xff;
+		int m = i & 0x007fffff;
+		int s = i & 0x80000000;
+		int v = add_sat(e, n);
+		v = clamp(v, 0, 0xff);
+		int mr = e == 0 | v == 0 | v == 0xff ? 0 : m;
+		int c = e == 0xff;
+		mr = c ? m : mr;
+		int er = c ? e : v;
+		er = e ? er : e;
+		return as_float( s | (er << 23) | mr );
+	}
+
+	/* supports denormal values */
+	const int multiplier = 24;
+	float val_f;
+	uint val_ui;
+	uint sign;
+	int exponent;
+	val_ui = as_uint(x);
+	sign = val_ui & 0x80000000;
+	val_ui = val_ui & 0x7fffffff;/* remove the sign bit */
+	int val_x = val_ui;
+
+	exponent = val_ui >> 23; /* get the exponent */
+	int dexp = exponent;
+
+	/* denormal support */
+	int fbh = 127 - (as_uint((float)(as_float(val_ui | 0x3f800000) - 1.0f)) >> 23);
+	int dexponent = 25 - fbh;
+	uint dval_ui = (( (val_ui << fbh) & 0x007fffff) | (dexponent << 23));
+	int ex = dexponent + n - multiplier;
+	dexponent = ex;
+	uint val = sign | (ex << 23) | (dval_ui & 0x007fffff);
+	int ex1 = dexponent + multiplier;
+	ex1 = -ex1 +25;
+	dval_ui = (((dval_ui & 0x007fffff )| 0x800000) >> ex1);
+	dval_ui = dexponent > 0 ? val :dval_ui;
+	dval_ui = dexponent > 254 ? 0x7f800000 :dval_ui;  /*overflow*/
+	dval_ui = dexponent < -multiplier ? 0 : dval_ui;  /*underflow*/
+	dval_ui = dval_ui | sign;
+	val_f = as_float(dval_ui);
+
+	exponent += n;
+
+	val = sign | (exponent << 23) | (val_ui & 0x007fffff);
+	ex1 = exponent + multiplier;
+	ex1 = -ex1 +25;
+	val_ui = (((val_ui & 0x007fffff )| 0x800000) >> ex1);
+	val_ui = exponent > 0 ? val :val_ui;
+	val_ui = exponent > 254 ? 0x7f800000 :val_ui;  /*overflow*/
+	val_ui = exponent < -multiplier ? 0 : val_ui;  /*underflow*/
+	val_ui = val_ui | sign;
+
+	val_ui = dexp == 0? dval_ui : val_ui;
+	val_f = as_float(val_ui);
+
+	val_f = isnan(x) | isinf(x) | val_x == 0 ? x : val_f;
+	return val_f;
+}
+
+#ifdef cl_khr_fp64
+
+#pragma OPENCL EXTENSION cl_khr_fp64 : enable
+
+_CLC_DEF _CLC_OVERLOAD double __clc_ldexp(double x, int n) {
+	long l = as_ulong(x);
+	int e = (l >> 52) & 0x7ff;
+	long s = l & 0x8000000000000000;
+
+	ulong ux = as_ulong(x * 0x1.0p+53);
+	int de = ((int)(ux >> 52) & 0x7ff) - 53;
+	int c = e == 0;
+	e = c ? de: e;
+
+	ux = c ? ux : l;
+
+	int v = e + n;
+	v = clamp(v, -0x7ff, 0x7ff);
+
+	ux &= ~EXPBITS_DP64;
+
+	double mr = as_double(ux | ((ulong)(v+53) << 52));
+	mr = mr * 0x1.0p-53;
+
+	mr = v > 0  ? as_double(ux | ((ulong)v << 52)) : mr;
+
+	mr = v == 0x7ff ? as_double(s | PINFBITPATT_DP64)  : mr;
+	mr = v < -53 ? as_double(s) : mr;
+
+	mr  = ((n == 0) | isinf(x) | (x == 0) ) ? x : mr;
+	return mr;
+}
+
```

Comment:
Should we also see the `half` version here, as in https://github.com/llvm/llvm-project/blob/96e6255e8b8b4e9e7bf0846df94dddcb79ced6f5/libclc/generic/lib/math/clc_ldexp.cl?

I don't think we're using these `__clc_ldexp` functions currently so I think it's better to stay as close to upstream as possible. Maybe we dropped the commit at some point, but we're in a bit of a libclc mess downstream right now.

Same goes for `clc_remquo` and the others you've added here.

Reply:
Oh I just noticed we have `clc_ldexp.cl` in `libclc/generic/libspirv/math/clc_ldexp.cl`. We direct `ldexp` through `__spirv_ocl_ldexp` through `__clc_ldexp`. That's confusing! We need to ensure that adding extra `__clc_ldexp` symbols in `generic/lib/math` isn't going to cause problems, or shadow the ones we "want" to link against.

Reply:
@frasercrmck  Thanks for the review. The resolution was done by coordinators who might not be familar with libclc. I have recovered the half version. Feel free to fix other conflicts resolution issues you found directly. Thanks!